ntoskrnl/config/cmnotify.c

```diff
@@ -12,25 +12,481 @@
+                CmpAcquireKcbLockShared(MasterKcb);
+
+                CmpNotifyPostBlock(PostBlock->MasterPostBlock);
+
+                CmpReleaseKcbLock(MasterKcb);
+            }
+        }
+    }
+}
+
+/**
+ * @brief
+ * Free the resources allocated for a NotifyBlock on a KeyBody.
+ * This is called when system is deleting a key body object, or it's closing a handle to a registry key.
+ *
+ * @param[in] LockHeld
+ * TRUE if calling while the KCB is already locked, FALSE if it not
+ */
+VOID NTAPI
+CmpFlushNotify(IN PCM_KEY_BODY KeyBody, IN BOOLEAN LockHeld)
+{
+    PLIST_ENTRY ListHead, NextEntry;
+    PCM_POST_BLOCK PostBlock;
+
+    /* Lock the KCB so no one would try to make changes
+     * to NotifyBlock while we are releasing its resources
+     */
+    if (!LockHeld)
+        CmpAcquireKcbLockExclusive(KeyBody->KeyControlBlock);
+
+    /* Enumerate PostBlocks */
+    ListHead = &(KeyBody->NotifyBlock->PostList);
+    NextEntry = ListHead->Flink;
+    while (NextEntry != ListHead)
+    {
+        PostBlock = CONTAINING_RECORD(NextEntry, CM_POST_BLOCK, NotifyList);
+        NextEntry = NextEntry->Flink;
+
+        /* This shouldn't happen, We don't assign subordinate NotifyBlocks to a KeyBody */
+        ASSERT(PostBlock->IsMasterPostBlock);
+
+        if (PostBlock->Event)
+            ObDereferenceObject(PostBlock->Event);
+
+        if (PostBlock->UserApc)
+            ExFreePoolWithTag(PostBlock->UserApc, TAG_CM);
+
+        RemoveEntryList(&(PostBlock->NotifyList));
+        CmpPostBlockFreeSubordinates(PostBlock->SubCount, PostBlock->SubNotifyBlocks);
+        ExFreePoolWithTag(PostBlock, TAG_CM);
+    }
+
+    /* Free the NotifyBlock */
+    RemoveEntryList(&(KeyBody->NotifyBlock->HiveList));
+    ExFreePoolWithTag(KeyBody->NotifyBlock, TAG_CM);
+    KeyBody->NotifyBlock = NULL;
+
+    /* Unlock the Kcb if we locked it previously */
+    if (!LockHeld)
+        CmpReleaseKcbLock(KeyBody->KeyControlBlock);
+}
+
+/**
+ * @brief
+ * Allocate a NotifyBlock for a registry key, and attach it to the hive.
+ * KeyBody->NotifyBlock should be set manually.
+ *
+ * @param[in] KeyBody
+ * The registry key to be watched for notifications
+ *
+ * @param[in] Filter
+ * The desired notification type, a bit flag of REG_CHANGE_FILTER_* enum values
+ */
+NTSTATUS
 NTAPI
-CmpReportNotify(IN PCM_KEY_CONTROL_BLOCK Kcb,
-                IN PHHIVE Hive,
-                IN HCELL_INDEX Cell,
-                IN ULONG Filter)
+CmpInsertNotifyBlock(
+    _In_ PCM_KEY_BODY KeyBody,
+    _In_ ULONG Filter,
+    _In_ BOOLEAN WatchTree,
+    _Out_ PCM_NOTIFY_BLOCK *Result)
 {
-    /* FIXME: TODO */
-    return;
+    PCM_NOTIFY_BLOCK NotifyBlock;
+    PCMHIVE Hive;
+
+    /* Allocate memory */
+    NotifyBlock = ExAllocatePoolWithTag(NonPagedPool, sizeof(CM_NOTIFY_BLOCK), TAG_CM);
+    if (!NotifyBlock)
+    {
+        *Result = NULL;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    RtlZeroMemory(NotifyBlock, sizeof(CM_NOTIFY_BLOCK));
```

Comment:
Use `ExAllocatePoolZero` for this and so you've saved yourself from having `RtlZeroMemory` in the code.